File: iosApp/iosApp/NearbyOffendersCard.swift
import SwiftUI

// ── Nearby Offenders Card ─────────────────────────────────────────────────────
// Ring fills from 0→progress and count increments from 0→count on appear.

struct NearbyOffendersCard: View {
    var count: Int = 247
    var progress: Double = 1.0
    var locationActive: Bool = true

    @Environment(\.twColors) private var colors
    @Environment(\.twTypography) private var typography

    @State private var animatedProgress: Double = 0
    @State private var animatedCount: Int = 0

    private let ringSize: CGFloat = 88
    private let ringStroke: CGFloat = 10
    private let duration: Double = 1.2

    var body: some View {
        HStack(alignment: .center, spacing: 16) {

            // ── Left: text info ───────────────────────────────────────────────
            VStack(alignment: .leading, spacing: 4) {
                Text("Nearby Offenders")
                    .font(typography.text2)
                    .foregroundColor(colors.secondaryText)

                HStack(alignment: .lastTextBaseline, spacing: 4) {
                    Text("\(animatedCount)")
                        .font(typography.h1)
                        .foregroundColor(colors.primaryText)
                        .contentTransition(.numericText())
                        .animation(.easeInOut(duration: duration), value: animatedCount)
                    Text("found")
                        .font(typography.text2)
                        .foregroundColor(colors.secondaryText)
                }

                Text("within 5 mile radius")
                    .font(typography.text2)
                    .foregroundColor(colors.secondaryText)
            }

            Spacer()

            // ── Right: donut ring + icon ──────────────────────────────────────
            ZStack {
                // Track ring
                Circle()
                    .stroke(colors.ringTrack,
                            style: StrokeStyle(lineWidth: ringStroke, lineCap: .round))
                    .frame(width: ringSize, height: ringSize)

                // Progress arc
                Circle()
                    .trim(from: 0, to: locationActive ? animatedProgress : 0)
                    .stroke(
                        locationActive ? colors.ringActive : colors.ringTrack,
                        style: StrokeStyle(lineWidth: ringStroke, lineCap: .round)
                    )
                    .frame(width: ringSize, height: ringSize)
                    .rotationEffect(.degrees(-90))
                    .animation(.easeInOut(duration: duration), value: animatedProgress)

                // Location pin icon
                Image(systemName: "location.fill")
                    .font(.system(size: 24, weight: .medium))
                    .foregroundColor(locationActive ? colors.ringActive : colors.ringTrack)
            }
        }
        .padding(20)
        .background(colors.cardBackground)
        .cornerRadius(16)
        .shadow(color: Color.black.opacity(0.12), radius: 8, x: 0, y: 3)
        .padding(.horizontal, 16)
        .padding(.top, 16)
        .onAppear {
            guard locationActive else { return }
            // Ring fills up
            withAnimation(.easeInOut(duration: duration)) {
                animatedProgress = progress
            }
            // Count increments step by step
            let steps = count
            let interval = duration / Double(max(steps, 1))
            for i in 1...max(steps, 1) {
                DispatchQueue.main.asyncAfter(deadline: .now() + interval * Double(i)) {
                    animatedCount = i
                }
            }
        }
    }
}

File: composeApp/src/androidMain/kotlin/com/jetbrains/spacetutorial/NearbyOffendersCard.kt
package com.jetbrains.spacetutorial

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.StrokeCap
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.unit.dp
import com.jetbrains.spacetutorial.texaswatch.theme.TexasWatchTheme
import kotlinx.coroutines.delay

private const val ANIM_DURATION_MS = 1200L

@Composable
fun NearbyOffendersCard(
    count: Int = 247,
    progress: Float = 1f,
    locationActive: Boolean = true,
) {
    val colors = TexasWatchTheme.colors
    val typography = TexasWatchTheme.typography

    val ringActive = colors.ringActive
    val ringTrack  = colors.ringTrack

    // ── Ring animation ────────────────────────────────────────────────────────
    var animatedTarget by remember { mutableFloatStateOf(0f) }
    LaunchedEffect(locationActive) {
        animatedTarget = if (locationActive) progress else 0f
    }
    val animatedProgress by animateFloatAsState(
        targetValue = animatedTarget,
        animationSpec = tween(durationMillis = ANIM_DURATION_MS.toInt()),
        label = "ringProgress",
    )

    // ── Count animation ───────────────────────────────────────────────────────
    var displayCount by remember { mutableIntStateOf(0) }
    LaunchedEffect(locationActive, count) {
        if (!locationActive) { displayCount = 0; return@LaunchedEffect }
        val steps = count
        val intervalMs = ANIM_DURATION_MS / steps.coerceAtLeast(1)
        for (i in 1..steps) {
            delay(intervalMs)
            displayCount = i
        }
    }

    val ringSize = 88.dp
    val strokeWidth = 10.dp

    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 16.dp, vertical = 8.dp)
            .shadow(elevation = 6.dp, shape = RoundedCornerShape(16.dp))
            .background(colors.cardBackground, RoundedCornerShape(16.dp))
            .padding(20.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        // ── Left: text ────────────────────────────────────────────────────────
        Column(
            verticalArrangement = Arrangement.spacedBy(4.dp),
            modifier = Modifier.weight(1f),
        ) {
            Text(
                text = "Nearby Offenders",
                style = typography.text2,
                color = colors.secondaryText,
            )
            Row(
                verticalAlignment = Alignment.Bottom,
                horizontalArrangement = Arrangement.spacedBy(4.dp),
            ) {
                Text(
                    text = "$displayCount",
                    style = typography.h1,
                    color = colors.primaryText,
                )
                Text(
                    text = "found",
                    style = typography.text2,
                    color = colors.secondaryText,
                )
            }
            Text(
                text = "within 5 mile radius",
                style = typography.text2,
                color = colors.secondaryText,
            )
        }

        Spacer(modifier = Modifier.size(16.dp))

        // ── Right: donut ring + icon ──────────────────────────────────────────
        Box(
            contentAlignment = Alignment.Center,
            modifier = Modifier
                .size(ringSize)
                .drawBehind {
                    val stroke = strokeWidth.toPx()
                    val inset  = stroke / 2f
                    val arcSize = Size(size.width - stroke, size.height - stroke)
                    val topLeft = Offset(inset, inset)

                    // Track
                    drawArc(
                        color      = ringTrack,
                        startAngle = -90f,
                        sweepAngle = 360f,
                        useCenter  = false,
                        topLeft    = topLeft,
                        size       = arcSize,
                        style      = Stroke(width = stroke, cap = StrokeCap.Round),
                    )

                    // Active arc
                    if (animatedProgress > 0f) {
                        drawArc(
                            color      = ringActive,
                            startAngle = -90f,
                            sweepAngle = 360f * animatedProgress,
                            useCenter  = false,
                            topLeft    = topLeft,
                            size       = arcSize,
                            style      = Stroke(width = stroke, cap = StrokeCap.Round),
                        )
                    }
                },
        ) {
            Icon(
                painter            = painterResource(R.drawable.location_28),
                contentDescription = "Location",
                tint               = if (locationActive) ringActive else ringTrack,
                modifier           = Modifier.size(24.dp),
            )
        }
    }
}
